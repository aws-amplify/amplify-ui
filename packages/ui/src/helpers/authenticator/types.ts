/**
 * This files provides types that describe general shape of
 * authenticator machine and its intepreter.
 */
import { Interpreter } from 'xstate';
import { AuthUser } from 'aws-amplify/auth';
import { HubCapsule } from 'aws-amplify/utils';

import { AuthContext, AuthEvent } from '../../machines/authenticator/types';

/**
 * Intefrace for `authMachine` machine interpreter
 *
 * TODO: tighten up anys here
 */
export type AuthInterpreter = Interpreter<
  AuthContext,
  any,
  AuthEvent,
  any,
  any
>;

/**
 * Function type for `send` in `authMachine`
 */
export type AuthMachineSend = AuthInterpreter['send'];

/**
 * Handles Amplify JS Auth hub events, by forwarding hub events as appropriate
 * xstate events.
 */
export type AuthMachineHubHandler = (
  data: HubCapsule<any, any>,
  service: AuthInterpreter,
  options?: {
    onSignIn?: (user: AuthUser) => void;
    onSignOut?: () => void;
  }
) => void;
