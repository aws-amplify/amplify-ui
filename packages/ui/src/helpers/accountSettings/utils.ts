import { Auth } from 'aws-amplify';

import { AmplifyUser } from '../../types';
import { getLogger } from '../utils';

const logger = getLogger('Auth');

type ChangePasswordInput = {
  user: AmplifyUser;
  currentPassword: string;
  newPassword: string;
};

export const changePassword = async ({
  user,
  currentPassword,
  newPassword,
}: ChangePasswordInput): Promise<void> => {
  try {
    logger.debug('calling Auth.changePassword');
    /**
     * Auth.changePassword returns `Promise<"SUCCESS">`. We're not interested
     * in its resolved string value, so we just return Promise.resolve() on success.
     */
    await Auth.changePassword(user, currentPassword, newPassword);
    logger.debug('Auth.changePassword was successful');
    return Promise.resolve();
  } catch (e) {
    logger.debug('Auth.changePassword failed with error', e);
    return Promise.reject(e);
  }
};

export const deleteUser = async () => {
  try {
    logger.debug('calling Auth.deleteUser');
    await Auth.deleteUser();
    logger.debug('Auth.deleteUser was successful');
    return Promise.resolve();
  } catch (e) {
    logger.debug('Auth.deleteUser failed with error', e);
    return Promise.reject(e);
  }
};
