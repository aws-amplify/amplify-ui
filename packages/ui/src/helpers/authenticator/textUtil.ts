import { SocialProvider } from '../../types';
import {
  ChallengeName,
  V5CodeDeliveryDetails,
} from '../../machines/authenticator/types';
import { translate, DefaultTexts } from '../../i18n';
import { AuthenticatorRoute } from './facade';

/**
 * ConfirmSignIn
 */
const getChallengeText = (challengeName?: ChallengeName): string => {
  switch (challengeName) {
    case 'SMS_MFA':
      return translate(DefaultTexts.CONFIRM_SMS);
    case 'SOFTWARE_TOKEN_MFA':
      return translate(DefaultTexts.CONFIRM_TOTP);
    default:
      return translate(DefaultTexts.CONFIRM_MFA_DEFAULT);
  }
};

/**
 * ConfirmSignUp
 */
const getDeliveryMessageText = (
  codeDeliveryDetails: V5CodeDeliveryDetails
): string => {
  const { DeliveryMedium, Destination } = codeDeliveryDetails ?? {};
  const isEmailMessage = DeliveryMedium === 'EMAIL';
  const isTextMessage = DeliveryMedium === 'SMS';

  const arrivalMessage = translate(DefaultTexts.CODE_ARRIVAL);

  if (!(isEmailMessage || isTextMessage)) {
    return `${translate(DefaultTexts.CODE_SENT)}. ${arrivalMessage}.`;
  }

  const instructionMessage = isEmailMessage
    ? translate(DefaultTexts.CODE_EMAILED)
    : translate(DefaultTexts.CODE_TEXTED);

  return `${instructionMessage} ${Destination}. ${arrivalMessage}.`;
};

const getDeliveryMethodText = (
  codeDeliveryDetails: V5CodeDeliveryDetails
): string => {
  const { DeliveryMedium } = codeDeliveryDetails ?? {};
  const isEmailMessage = DeliveryMedium === 'EMAIL';
  const isTextMessage = DeliveryMedium === 'SMS';

  if (!isEmailMessage && isTextMessage) {
    return translate(DefaultTexts.WE_SENT_CODE);
  }
  return isEmailMessage
    ? translate(DefaultTexts.WE_EMAILED)
    : translate(DefaultTexts.WE_TEXTED);
};

/**
 * FederatedSignIn
 */
const providerNameMap: Record<SocialProvider, string> = {
  amazon: 'Amazon',
  apple: 'Apple',
  facebook: 'Facebook',
  google: 'Google',
};

const getSignInWithFederationText = (
  route: AuthenticatorRoute,
  provider: SocialProvider
) => {
  const isSignIn = route === 'signIn';

  return translate(
    `Sign ${isSignIn ? 'In' : 'Up'} with ${providerNameMap[provider]}`
  );
};

export const authenticatorTextUtil = {
  /** Shared */
  getBackToSignInText: () => translate(DefaultTexts.BACK_SIGN_IN),
  getChangePasswordText: () => translate(DefaultTexts.CHANGE_PASSWORD),
  getChangingText: () => translate(DefaultTexts.CHANGING_PASSWORD),
  getConfirmText: () => translate(DefaultTexts.CONFIRM),
  getConfirmingText: () => translate(DefaultTexts.CONFIRMING),
  getCopyText: () => translate(DefaultTexts.UPPERCASE_COPY),
  getHidePasswordText: () => translate(DefaultTexts.HIDE_PASSWORD),
  getLoadingText: () => translate(DefaultTexts.LOADING),
  getOrText: () => translate(DefaultTexts.OR),
  getResendCodeText: () => translate(DefaultTexts.RESEND_CODE),
  getSendCodeText: () => translate(DefaultTexts.SEND_CODE),
  getSendingText: () => translate(DefaultTexts.SENDING),
  getShowPasswordText: () => translate(DefaultTexts.SHOW_PASSWORD),
  getSubmitText: () => translate(DefaultTexts.SUBMIT),
  getSubmittingText: () => translate(DefaultTexts.SUBMITTING),

  /** SignInSignUpTabs */
  getSignInTabText: () => translate(DefaultTexts.SIGN_IN_TAB),
  getSignUpTabText: () => translate(DefaultTexts.CREATE_ACCOUNT),

  /** SignIn */
  getForgotPasswordText: (shortVersion?: boolean) =>
    translate(
      shortVersion
        ? DefaultTexts.FORGOT_PASSWORD
        : DefaultTexts.FORGOT_YOUR_PASSWORD
    ),
  getSigningInText: () => translate(DefaultTexts.SIGNING_IN_BUTTON),
  getSignInText: () => translate(DefaultTexts.SIGN_IN_BUTTON),

  /** SignUp */
  getCreatingAccountText: () => translate(DefaultTexts.CREATING_ACCOUNT),
  getCreateAccountText: () => translate(DefaultTexts.CREATE_ACCOUNT),

  /** ConfirmSignUp */
  getDeliveryMessageText,
  getDeliveryMethodText,

  /** ConfirmSignIn */
  getChallengeText,

  /** ForgotPassword */
  getResetYourPasswordText: () => translate(DefaultTexts.RESET_PASSWORD),

  /** SetupTotp */
  getSetupTotpText: () => translate(DefaultTexts.SETUP_TOTP),
  // TODO: add defaultText for below
  getSetupTotpInstructionsText: () =>
    translate(
      'Copy and paste the secret key below into an authenticator app and then enter the code in the text field below.'
    ),
  // TODO: add defaultText for "COPIED"
  getCopiedText: () => translate('COPIED'),

  /** FederatedSignIn */
  getSignInWithFederationText,

  /** VerifyUser */
  getSkipText: () => translate(DefaultTexts.SKIP),
  getVerifyText: () => translate(DefaultTexts.VERIFY),
  getVerifyContactText: () => translate(DefaultTexts.VERIFY_CONTACT),
  getAccountRecoveryInfoText: () => translate(DefaultTexts.VERIFY_HEADING),

  /** Validations */
  // TODO: add defaultText
  getInvalidEmailText: () => translate('Please enter a valid email'),
  // TODO: add defaultText
  getRequiredFieldText: () => translate('This field is required'),
} as const; // using `as const` so that keys are strongly typed
