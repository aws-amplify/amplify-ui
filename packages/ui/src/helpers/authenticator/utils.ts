/**
 * This file contains general helpers that state machine or authenticator
 * implementations can use.
 */

import { Hub } from 'aws-amplify';
import { AuthInterpreter, HubHandler } from '../../types';
import { ALLOWED_SPECIAL_CHARACTERS } from './constants';

// replaces all characters in a string with '*', except for the first and last char
export const censorAllButFirstAndLast = (value: string): string => {
  const split = value.trim().split('');
  for (let i = 0; i < split.length; i++) {
    if (i > 0 && i < split.length - 1) {
      split[i] = '*';
    }
  }

  return split.join('');
};

// censors all but the last four characters of a phone number
export const censorPhoneNumber = (val: string): string => {
  if (val.length < 4) {
    return val;
  }

  const split = val.split('');
  for (let i = 0; i < split.length - 4; i++) {
    split[i] = '*';
  }

  return split.join('');
};

export const defaultAuthHubHandler: HubHandler = (data, service) => {
  const { send } = service;
  const state = service.getSnapshot(); // this is just a getter and is not expensive

  switch (data.payload.event) {
    // TODO: We can add more cases here, according to
    // https://docs.amplify.aws/lib/auth/auth-events/q/platform/js/
    case 'tokenRefresh':
      if (state.matches('authenticated.idle')) {
        send('TOKEN_REFRESH');
      }
      break;
    case 'signOut':
    case 'tokenRefresh_failure':
      if (state.matches('authenticated.idle')) {
        send('SIGN_OUT');
      }
      break;
    default:
      break;
  }
};

/**
 * Listens to external auth Hub events and sends corresponding event to
 * the `authService` of interest
 *
 * @param send - `send` function associated with the `authService` of interest
 *
 * @returns function that unsubscribes to the hub evenmt
 */
export const listenToAuthHub = (
  service: AuthInterpreter,
  handler: HubHandler = defaultAuthHubHandler
) => {
  return Hub.listen(
    'auth',
    (data) => {
      handler(data, service);
    },
    'authenticator-hub-handler'
  );
};

export const hasSpecialChars = (password: string) =>
  ALLOWED_SPECIAL_CHARACTERS.some((char) => password.includes(char));
