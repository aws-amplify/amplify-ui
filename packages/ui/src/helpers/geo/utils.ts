import {
  Category,
  GeoAction,
  SetCustomUserAgentInput,
} from '@aws-amplify/core/internals/utils';

export const geoDataPlaneState: SetCustomUserAgentInput = {
  category: Category.Geo,
  apis: [GeoAction.None],
  additionalDetails: [['component', 'geo']],
};
