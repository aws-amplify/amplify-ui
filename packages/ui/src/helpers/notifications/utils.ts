import {
  Category,
  InAppMessagingAction,
  SetCustomUserAgentInput,
} from '@aws-amplify/core/internals/utils';

export const notificationsDataPlaneState: SetCustomUserAgentInput = {
  category: Category.InAppMessaging,
  apis: [InAppMessagingAction.SyncMessages],
  additionalDetails: [['component', 'inappmessaging']],
};
