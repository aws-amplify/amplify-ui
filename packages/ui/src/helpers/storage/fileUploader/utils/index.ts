export {
  uploadFile,
  humanFileSize,
  checkMaxSize,
  returnAcceptedFiles,
  isValidExtension,
} from './uploader';
