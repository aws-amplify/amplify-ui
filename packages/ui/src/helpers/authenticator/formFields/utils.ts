import { translate } from '../../../i18n/translations';
import { FormFields, FormFieldsArray } from '../../../types';
import { AuthMachineState } from '../../../machines/authenticator/types';

export const getPrimaryAlias = (state: AuthMachineState) => {
  const loginMechanisms = state?.context.config?.loginMechanisms;
  /**
   * @migration this is where we grab only the first index of `aws_cognito_username_attributes`
   */
  const [primaryAlias] = loginMechanisms ?? ['username'];
  return primaryAlias;
};

/** Applies translations to label and placeholder */
export const applyTranslation = (formFields: FormFields): FormFields => {
  const newFormFields = { ...formFields };
  for (const [name, options] of Object.entries(formFields)) {
    const { label, placeholder } = options;

    newFormFields[name] = {
      ...options,
      label: label ? translate<string>(label) : undefined,
      placeholder: placeholder ? translate<string>(placeholder) : undefined,
    };
  }
  return newFormFields;
};

/** Sorts formFields according to their `order`.  */
export const sortFormFields = (formFields: FormFields): FormFieldsArray => {
  return Object.entries(formFields)
    .sort((a, b) => {
      const orderA = a[1].order || Number.MAX_VALUE;
      const orderB = b[1].order || Number.MAX_VALUE;
      return orderA - orderB;
    })
    .filter((formFieldEntry) => formFieldEntry[1] !== undefined);
};
