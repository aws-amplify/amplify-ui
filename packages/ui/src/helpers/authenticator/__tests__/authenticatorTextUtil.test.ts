import { I18n } from 'aws-amplify/utils';
import { SocialProvider } from '../../../types';
import {
  ChallengeName,
  V5CodeDeliveryDetails,
} from '../../../machines/authenticator/types';
import { authenticatorTextUtil } from '../textUtil';

const AUTH_CHALLENGE_NAMES: ChallengeName[] = ['SMS_MFA', 'SOFTWARE_TOKEN_MFA'];

const SOCIAL_PROVIDERS: SocialProvider[] = [
  'apple',
  'amazon',
  'facebook',
  'google',
];

const CODE_DELIVERY_DETAILS: Array<V5CodeDeliveryDetails['DeliveryMedium']> = [
  'SMS',
  'EMAIL',
];

I18n.setLanguage('en-us');
describe('authenticatorTextUtil', () => {
  beforeEach(() => {
    // reset translations
    I18n.putVocabulariesForLanguage('en-us', { 'Sign in': 'Sign in' });
  });

  it('should reflect i18n translation', () => {
    const { getSignInText } = authenticatorTextUtil;
    expect(getSignInText()).toBe('Sign in');
    I18n.putVocabulariesForLanguage('en-us', { 'Sign in': 'Log in' });
    expect(getSignInText()).toBe('Log in');
  });

  it.each(SOCIAL_PROVIDERS)(
    'getSignInWithFederationText with %s in sign in route renders text as expected',
    (provider) => {
      const { getSignInWithFederationText } = authenticatorTextUtil;
      expect(getSignInWithFederationText('signIn', provider)).toMatchSnapshot();
    }
  );

  it.each(SOCIAL_PROVIDERS)(
    'getSignInWithFederationText with %s in sign up route returns the expected text',
    (provider) => {
      const { getSignInWithFederationText } = authenticatorTextUtil;
      expect(getSignInWithFederationText('signUp', provider)).toMatchSnapshot();
    }
  );

  it.each(CODE_DELIVERY_DETAILS)(
    'getDeliveryMethodText for %s returns the expected text',
    (deliveryMethod) => {
      const { getDeliveryMethodText } = authenticatorTextUtil;

      const codeDeliveryDetail = {
        DeliveryMedium: deliveryMethod,
      } as unknown as V5CodeDeliveryDetails;

      expect(getDeliveryMethodText(codeDeliveryDetail)).toMatchSnapshot();
    }
  );

  it.each(CODE_DELIVERY_DETAILS)(
    'getDeliveryMessageText for %s returns the expected text',
    (deliveryMethod) => {
      const { getDeliveryMessageText } = authenticatorTextUtil;

      const codeDeliveryDetail = {
        DeliveryMedium: deliveryMethod,
      } as unknown as V5CodeDeliveryDetails;

      expect(getDeliveryMessageText(codeDeliveryDetail)).toMatchSnapshot();
    }
  );

  it.each(AUTH_CHALLENGE_NAMES)(
    'getChallengeText for %s returns the expected text',
    (challengeName) => {
      const { getChallengeText } = authenticatorTextUtil;
      expect(getChallengeText(challengeName)).toMatchSnapshot();
    }
  );
});
