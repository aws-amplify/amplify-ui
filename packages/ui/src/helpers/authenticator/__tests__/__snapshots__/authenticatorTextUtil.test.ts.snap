// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticatorTextUtil getChallengeText for SMS_MFA returns the expected text 1`] = `"Confirm SMS Code"`;

exports[`authenticatorTextUtil getChallengeText for SOFTWARE_TOKEN_MFA returns the expected text 1`] = `"Confirm TOTP Code"`;

exports[`authenticatorTextUtil getDeliveryMessageText for EMAIL returns the expected text 1`] = `"Your code is on the way. To log in, enter the code we emailed to undefined. It may take a minute to arrive."`;

exports[`authenticatorTextUtil getDeliveryMessageText for SMS returns the expected text 1`] = `"Your code is on the way. To log in, enter the code we texted to undefined. It may take a minute to arrive."`;

exports[`authenticatorTextUtil getDeliveryMethodText for EMAIL returns the expected text 1`] = `"We Emailed You"`;

exports[`authenticatorTextUtil getDeliveryMethodText for SMS returns the expected text 1`] = `"We Sent A Code"`;

exports[`authenticatorTextUtil getSignInWithFederationText with amazon in sign in route renders text as expected 1`] = `"Sign In with Amazon"`;

exports[`authenticatorTextUtil getSignInWithFederationText with amazon in sign up route returns the expected text 1`] = `"Sign Up with Amazon"`;

exports[`authenticatorTextUtil getSignInWithFederationText with apple in sign in route renders text as expected 1`] = `"Sign In with Apple"`;

exports[`authenticatorTextUtil getSignInWithFederationText with apple in sign up route returns the expected text 1`] = `"Sign Up with Apple"`;

exports[`authenticatorTextUtil getSignInWithFederationText with facebook in sign in route renders text as expected 1`] = `"Sign In with Facebook"`;

exports[`authenticatorTextUtil getSignInWithFederationText with facebook in sign up route returns the expected text 1`] = `"Sign Up with Facebook"`;

exports[`authenticatorTextUtil getSignInWithFederationText with google in sign in route renders text as expected 1`] = `"Sign In with Google"`;

exports[`authenticatorTextUtil getSignInWithFederationText with google in sign up route returns the expected text 1`] = `"Sign Up with Google"`;
